modules/gmc-agent/src/main/java/org/gridgain/action/SessionRegistry.java

```diff
@@ -0,0 +1,127 @@
+ * limitations under the License.
+ */
+
+package org.gridgain.action;
+
+import org.apache.ignite.IgniteAuthenticationException;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.IgniteSystemProperties;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.gmc.ManagementConfiguration;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.gridgain.agent.AgentUtils;
+
+import java.time.Duration;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_REST_SECURITY_TOKEN_TIMEOUT;
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_REST_SESSION_TIMEOUT;
+
+/**
+ * Session registry.
+ */
+public class SessionRegistry {
+    /** Instance. */
+    private static volatile SessionRegistry instance;
+
+    /** Context. */
+    private final GridKernalContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** SessionId-Session map. */
+    private final ConcurrentMap<UUID, Session> sesId2Ses = new ConcurrentHashMap<>();
+
+    /** Session time to live. */
+    private final Duration sesTtl;
+
+    /** Interval to invalidate session tokens. */
+    private final Duration sesTokTtl;
+
+    /**
+     * @param ctx Context.
+     */
+    private SessionRegistry(GridKernalContext ctx) {
+        this.ctx = ctx;
+        this.log = ctx.log(SessionRegistry.class);
+
+        ManagementConfiguration cfg = ctx.gmc().configuration();
+
+        sesTtl = Duration.ofMillis(cfg.getSessionTimeout());
+        sesTokTtl = Duration.ofMillis(cfg.getSessionExpirationTimeout());
+    }
+
+    /**
+     * @param ctx Context.
+     */
+    public static SessionRegistry getInstance(GridKernalContext ctx) {
+        SessionRegistry locInstance = instance;
+
+        if (locInstance == null) {
+            synchronized (SessionRegistry.class) {
+                locInstance = instance;
+
+                if (locInstance == null)
+                    instance = locInstance = new SessionRegistry(ctx);
+            }
+        }
+
+        return locInstance;
+    }
+
+    /**
+     * @param ses Session.
+     */
+    public void saveSession(Session ses) {
+        sesId2Ses.put(ses.id(), ses);
+    }
+
+    /**
+     * @param sesId Session ID.
+     * @return Not null session.
+     * @throws IgniteAuthenticationException If failed.
+     */
+    public Session getSession(UUID sesId) throws IgniteCheckedException {
+        if (sesId == null)
+            throw new IgniteAuthenticationException("Session id can't be null, please authenticate and provide session id");
+
+        Session ses = sesId2Ses.get(sesId);
+
+        if (ses == null)
+            throw new IgniteAuthenticationException("Failed to retrieve session by session id, [sessionId=" + sesId + "]");
```

Comment:
Lets use message: "Session not found for ID: " + sesId

Reply:
Fix.